File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A01_train_test_y_mse/R/A01_train_test_y_mse.R
# ############################################################################
# A1. PARTICIÓN
# Tema A01_train_test_y_mse — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript A01_train_test_y_mse.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("A1. PARTICIÓN\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# A1. PARTICIÓN TRAIN/TEST Y MSE
# ----------------------------------------------------------------------------
# QUÉ HACE: parte los datos en dos. Con una parte entrenas, con la otra
# calificas. Es como estudiar con unos ejercicios y presentar el examen con
# otros: si te calificas con los mismos que estudiaste, la nota no dice nada.
#
# POR QUÉ: el error en TRAIN siempre baja si haces el modelo más flexible.
# Un modelo que memoriza tiene error de train = 0 y no sirve. El que importa
# es el MSE de TEST.
#
#   MSE_test = (1/n_test) * suma( (y_pred - y_real)^2 )
#
# PARÁMETROS
#   proporción train : entre 0.50 y 0.90; lo típico es 0.70 o 0.67
#   set.seed(n)      : PONLO SIEMPRE, para que el resultado se repita
# ============================================================================

# La partición ya está hecha arriba. Manualmente sería:
set.seed(0)
idx2   <- sample(seq_len(nrow(X)), size = floor(0.70*nrow(X)))   # 70% train
Xtr <- X[idx2,]; Xte <- X[-idx2,]; ytr <- y[idx2]; yte <- y[-idx2]

# Cómo calcular el MSE de cualquier modelo:
modelo <- lm(ytrain ~ ., data = data.frame(Xtrain, ytrain=ytrain))  # 1. entrenar
pred   <- predict(modelo, newdata = data.frame(Xtest))              # 2. predecir
mse    <- mean((ytest - pred)^2)                                    # 3. comparar

cat(sprintf("MSE en test : %.3f\n", mse))
cat(sprintf("RMSE en test: %.3f  (en las unidades de y)\n", sqrt(mse)))

# GRÁFICA estándar: predicho vs. real
plot(ytest, pred, pch=16, col=rgb(0,0,1,.5),
     xlab="y real (test)", ylab="y predicho",
     main=sprintf("Predicho vs. real — RMSE = %.2f", sqrt(mse)))
abline(0, 1, col="red", lwd=2, lty=2)
# Entre más pegados los puntos a la línea roja, mejor el modelo.

cat("\nListo. Revisa la carpeta figuras/\n")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A01_train_test_y_mse/python/A01_train_test_y_mse.py
"""
A1. Partición train / test y el MSE
Tema A01_train_test_y_mse — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python A01_train_test_y_mse.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("A1. Partición train / test y el MSE")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ A1. Partición train/test ═══

# Ya está hecha arriba, pero así se hace manualmente si necesitas otra:
X_tr, X_te, y_tr, y_te = train_test_split(
    X, y,
    test_size=0.30,        # 30% para test
    random_state=0,        # semilla fija -> reproducible
    # stratify=y,          # ⬅ DESCOMENTA si es CLASIFICACIÓN
)
print(f"train: {len(y_tr)} obs.  |  test: {len(y_te)} obs.")

# ── Cómo calcular el MSE de cualquier modelo ────────────────────────────
modelo = ModeloLineal().fit(Xs_train, y_train)      # 1. entrenas con TRAIN
pred = modelo.predict(Xs_test)                      # 2. predices en TEST
mse = mean_squared_error(y_test, pred)              # 3. comparas con el y real

print(f"\nMSE en test : {mse:.3f}")
print(f"RMSE en test: {np.sqrt(mse):.3f}   <- en las mismas unidades que y")

# También a mano, por si piden la fórmula:
print(f"MSE a mano  : {np.mean((pred - y_test)**2):.3f}")

# ── Gráfica estándar: predicho vs. real (solo tiene sentido en regresión) ──
if ES_CLASIFICACION:
    print("\n(En clasificación mira la matriz de confusión, sección A4.)")
plt.scatter(y_test, pred, alpha=.6)
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], "r--", lw=2)
plt.xlabel("y real (test)"); plt.ylabel("y predicho")
plt.title(f"Predicho vs. real — RMSE = {np.sqrt(mse):.2f}")
plt.show()
print("Entre más pegados los puntos a la línea roja, mejor el modelo.")

print("\nListo. Revisa la carpeta figuras/")
